File: android/src/main/kotlin/dev/brewkits/native_workmanager/workers/FileSystemWorker.kt
package dev.brewkits.native_workmanager.workers

import android.content.Context
import androidx.work.WorkerParameters
import dev.brewkits.kmpworkmanager.background.domain.AndroidWorker
import dev.brewkits.kmpworkmanager.background.domain.WorkerResult
import dev.brewkits.native_workmanager.workers.utils.ProgressReporter
import dev.brewkits.native_workmanager.workers.utils.SecurityValidator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.buildJsonArray
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption

/**
 * Built-in worker: File system operations
 *
 * Supports copy, move, delete, list, and mkdir operations for pure-native task chains.
 *
 * **Configuration JSON:**
 * ```json
 * // Copy operation
 * {
 *   "operation": "copy",
 *   "sourcePath": "/path/to/source",
 *   "destinationPath": "/path/to/destination",
 *   "overwrite": false,
 *   "recursive": true
 * }
 *
 * // Move operation
 * {
 *   "operation": "move",
 *   "sourcePath": "/path/to/source",
 *   "destinationPath": "/path/to/destination",
 *   "overwrite": false
 * }
 *
 * // Delete operation
 * {
 *   "operation": "delete",
 *   "path": "/path/to/file",
 *   "recursive": false
 * }
 *
 * // List operation
 * {
 *   "operation": "list",
 *   "path": "/path/to/directory",
 *   "pattern": "*.jpg",
 *   "recursive": false
 * }
 *
 * // Mkdir operation
 * {
 *   "operation": "mkdir",
 *   "path": "/path/to/new/directory",
 *   "createParents": true
 * }
 * ```
 */
class FileSystemWorker : AndroidWorker {

    override suspend fun doWork(input: String?, env: dev.brewkits.kmpworkmanager.background.domain.WorkerEnvironment): WorkerResult = withContext(Dispatchers.IO) {
        try {
            if (input.isNullOrEmpty()) {
                return@withContext WorkerResult.Failure("Input JSON is required")
            }

            val json = JSONObject(input)
            val operation = json.getString("operation")

            when (operation) {
                "copy" -> handleCopy(json)
                "move" -> handleMove(json)
                "delete" -> handleDelete(json)
                "list" -> handleList(json)
                "mkdir" -> handleMkdir(json)
                else -> WorkerResult.Failure("Unknown operation: $operation")
            }
        } catch (e: Exception) {
            WorkerResult.Failure("FileSystem operation failed: ${e.message}")
        }
    }

    private fun handleCopy(json: JSONObject): WorkerResult {
        val sourcePath = json.getString("sourcePath")
        val destinationPath = json.getString("destinationPath")
        val overwrite = json.optBoolean("overwrite", false)
        val recursive = json.optBoolean("recursive", true)

        // H4: Validate paths BEFORE any existence checks to avoid information
        // disclosure (different error messages revealing whether a restricted
        // path exists before the security check fires).
        if (!SecurityValidator.validateFilePathSafe(sourcePath) || !SecurityValidator.validateFilePathSafe(destinationPath)) {
            return WorkerResult.Failure("Invalid or unsafe path")
        }

        val sourceFile = File(sourcePath)
        if (!sourceFile.exists()) {
            return WorkerResult.Failure("Source not found: $sourcePath")
        }

        val destFile = File(destinationPath)

        // Check if destination exists
        if (destFile.exists() && !overwrite) {
            return WorkerResult.Failure("Destination already exists: $destinationPath (set overwrite=true to replace)")
        }

        return try {
            val copiedFiles = if (sourceFile.isDirectory) {
                if (!recursive) {
                    return WorkerResult.Failure("Source is a directory, set recursive=true to copy")
                }
                copyDirectory(sourceFile, destFile, overwrite)
            } else {
                copyFile(sourceFile, destFile, overwrite)
                listOf(destFile)
            }

            val totalSize = copiedFiles.sumOf { it.length() }

            WorkerResult.Success(
                message = "Copied ${copiedFiles.size} file(s)",
                data = buildJsonObject {
                    put("operation", "copy")
                    put("sourcePath", sourcePath)
                    put("destinationPath", destinationPath)
                    put("fileCount", copiedFiles.size)
                    put("totalSize", totalSize)
                    put("files", buildJsonArray { copiedFiles.forEach { add(kotlinx.serialization.json.JsonPrimitive(it.absolutePath)) } })
                }
            )
        } catch (e: IOException) {
            WorkerResult.Failure("Copy failed: ${e.message}")
        }
    }

    private fun handleMove(json: JSONObject): WorkerResult {
        val sourcePath = json.getString("sourcePath")
        val destinationPath = json.getString("destinationPath")
        val overwrite = json.optBoolean("overwrite", false)

        // H4: Validate paths before existence check (avoid information disclosure).
        if (!SecurityValidator.validateFilePathSafe(sourcePath) || !SecurityValidator.validateFilePathSafe(destinationPath)) {
            return WorkerResult.Failure("Invalid or unsafe path")
        }

        val sourceFile = File(sourcePath)
        if (!sourceFile.exists()) {
            return WorkerResult.Failure("Source not found: $sourcePath")
        }

        val destFile = File(destinationPath)

        if (destFile.exists() && !overwrite) {
            return WorkerResult.Failure("Destination already exists: $destinationPath (set overwrite=true to replace)")
        }

        return try {
            // FS-H-006: Check mkdirs return value
            val destParent = destFile.parentFile
            if (destParent != null && !destParent.exists() && !destParent.mkdirs()) {
                return WorkerResult.Failure("Failed to create destination parent directory")
            }

            // Delete destination if overwriting
            if (destFile.exists() && overwrite) {
                destFile.deleteRecursively()
            }

            // Attempt atomic move first
            val moved = sourceFile.renameTo(destFile)

            if (!moved) {
                // Fallback: copy + delete
                if (sourceFile.isDirectory) {
                    copyDirectory(sourceFile, destFile, overwrite)
                } else {
                    copyFile(sourceFile, destFile, overwrite)
                }
                // FS-C-002: rollback copy if source delete fails
                val deleted = sourceFile.deleteRecursively()
                if (!deleted) {
                    destFile.deleteRecursively()
                    return WorkerResult.Failure("Move failed: could not delete source after copy")
                }
            }

            val fileCount = if (destFile.isDirectory) {
                destFile.walkTopDown().count { it.isFile }
            } else 1

            WorkerResult.Success(
                message = "Moved $fileCount file(s)",
                data = buildJsonObject {
                    put("operation", "move")
                    put("sourcePath", sourcePath)
                    put("destinationPath", destinationPath)
                    put("fileCount", fileCount)
                }
            )
        } catch (e: IOException) {
            WorkerResult.Failure("Move failed: ${e.message}")
        }
    }

    private fun handleDelete(json: JSONObject): WorkerResult {
        val path = json.getString("path")
        val recursive = json.optBoolean("recursive", false)

        // H3: Validate path with canonical resolution before any file operations.
        if (!SecurityValidator.validateFilePathSafe(path)) {
            return WorkerResult.Failure("Invalid or unsafe path")
        }

        val file = File(path)
        if (!file.exists()) {
            return WorkerResult.Failure("Path not found: $path")
        }

        // Additional safety check: prevent accidental deletion of root-level directories.
        val dangerousPaths = listOf("/", "/system", "/data", "/storage/emulated/0")
        if (dangerousPaths.any { file.canonicalPath == File(it).canonicalPath }) {
            return WorkerResult.Failure("Cannot delete protected path: $path")
        }

        return try {
            val fileCount = if (file.isDirectory) {
                if (!recursive) {
                    return WorkerResult.Failure("Path is a directory, set recursive=true to delete")
                }
                file.walkTopDown().count { it.isFile }
            } else 1

            val deleted = if (file.isDirectory && recursive) {
                file.deleteRecursively()
            } else {
                file.delete()
            }

            // FS-H-007: deleteRecursively returns false on partial deletion; check if
            // target still exists to distinguish outright failure from partial success.
            if (deleted || !file.exists()) {
                WorkerResult.Success(
                    message = "Deleted $fileCount file(s)",
                    data = buildJsonObject {
                        put("operation", "delete")
                        put("path", path)
                        put("fileCount", fileCount)
                    }
                )
            } else {
                WorkerResult.Failure("Failed to delete (some files may remain): $path")
            }
        } catch (e: IOException) {
            WorkerResult.Failure("Delete failed: ${e.message}")
        }
    }

    private fun handleList(json: JSONObject): WorkerResult {
        val path = json.getString("path")
        val pattern = json.optString("pattern").takeIf { it.isNotEmpty() }
        val recursive = json.optBoolean("recursive", false)

        // H3: Validate path before listing — missing in original code.
        if (!SecurityValidator.validateFilePathSafe(path)) {
            return WorkerResult.Failure("Invalid or unsafe path")
        }

        val directory = File(path)
        if (!directory.exists()) return WorkerResult.Failure("Path not found: $path")
        if (!directory.isDirectory) return WorkerResult.Failure("Path is not a directory: $path")

        return try {
            // Use sequence to avoid loading everything into memory (prevent OOM)
            val fileSequence = if (recursive) {
                directory.walkTopDown().filter { it.isFile }
            } else {
                directory.listFiles()?.asSequence()?.filter { it.isFile } ?: emptySequence()
            }

            // Apply pattern filter efficiently
            val regex = pattern?.let {
                // Fix Regex Injection — properly quote the entire pattern to escape all meta-characters,
                // then unquote the specific wildcards (* and ?) we want to support.
                val escaped = java.util.regex.Pattern.quote(it)
                    .replace("\\*", ".*")
                    .replace("\\?", ".")
                try {
                    "^$escaped$".toRegex(RegexOption.IGNORE_CASE)
                } catch (e: Exception) {
                    null // Fallback if pattern is invalid
                }
            }

            val filteredFiles = if (regex != null) {
                fileSequence.filter { regex.matches(it.name) }
            } else {
                fileSequence
            }

            // Limit result size for stability (e.g., max 1000 items in response)
            val resultList = filteredFiles.take(1000).toList()

            WorkerResult.Success(
                message = "Found ${resultList.size} file(s)",
                data = buildJsonObject {
                    put("operation", "list")
                    put("files", buildJsonArray {
                        resultList.forEach { file ->
                            add(buildJsonObject {
                                put("path", file.absolutePath)
                                put("name", file.name)
                                put("size", file.length())
                                put("lastModified", file.lastModified())   // FS-M-004
                                put("isDirectory", file.isDirectory)       // FS-L-003
                            })
                        }
                    })
                }
            )
        } catch (e: Exception) {
            WorkerResult.Failure("List failed: ${e.message}")
        }
    }

    private fun handleMkdir(json: JSONObject): WorkerResult {
        val path = json.getString("path")
        val createParents = json.optBoolean("createParents", true)

        // FS-M-005: Security check BEFORE exists() to avoid information disclosure
        if (!SecurityValidator.validateFilePathSafe(path)) {
            return WorkerResult.Failure("Invalid or unsafe path")
        }

        val directory = File(path)

        if (directory.exists()) {
            return if (directory.isDirectory) {
                WorkerResult.Success(
                    message = "Directory already exists",
                    data = buildJsonObject {
                        put("operation", "mkdir")
                        put("path", path)
                        put("created", false)
                    }
                )
            } else {
                WorkerResult.Failure("Path exists but is not a directory: $path")
            }
        }

        return try {
            val created = if (createParents) {
                directory.mkdirs()
            } else {
                directory.mkdir()
            }

            if (created) {
                WorkerResult.Success(
                    message = "Directory created",
                    data = buildJsonObject {
                        put("operation", "mkdir")
                        put("path", path)
                        put("created", true)
                    }
                )
            } else {
                WorkerResult.Failure("Failed to create directory: $path")
            }
        } catch (e: IOException) {
            WorkerResult.Failure("Mkdir failed: ${e.message}")
        }
    }

    private fun copyFile(source: File, destination: File, overwrite: Boolean) {
        destination.parentFile?.mkdirs()

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            val copyOption = if (overwrite) {
                StandardCopyOption.REPLACE_EXISTING
            } else {
                StandardCopyOption.COPY_ATTRIBUTES
            }
            Files.copy(source.toPath(), destination.toPath(), copyOption)
        } else {
            // Fallback for older Android versions
            source.copyTo(destination, overwrite = overwrite)
        }
    }

    private fun copyDirectory(source: File, destination: File, overwrite: Boolean): List<File> {
        val copiedFiles = mutableListOf<File>()

        source.walkTopDown().forEach { file ->
            val relativePath = file.relativeTo(source).path
            val destFile = File(destination, relativePath)

            if (file.isDirectory) {
                destFile.mkdirs()
            } else {
                copyFile(file, destFile, overwrite)
                copiedFiles.add(destFile)
            }
        }

        return copiedFiles
    }
}

File: ios/native_workmanager/Sources/native_workmanager/workers/FileSystemWorker.swift
import Foundation
import KMPWorkManager

/// Built-in worker: File system operations
///
/// Supports copy, move, delete, list, and mkdir operations for pure-native task chains.
///
/// **Configuration JSON:**
/// ```json
/// // Copy operation
/// {
///   "operation": "copy",
///   "sourcePath": "/path/to/source",
///   "destinationPath": "/path/to/destination",
///   "overwrite": false,
///   "recursive": true
/// }
///
/// // Move operation
/// {
///   "operation": "move",
///   "sourcePath": "/path/to/source",
///   "destinationPath": "/path/to/destination",
///   "overwrite": false
/// }
///
/// // Delete operation
/// {
///   "operation": "delete",
///   "path": "/path/to/file",
///   "recursive": false
/// }
///
/// // List operation
/// {
///   "operation": "list",
///   "path": "/path/to/directory",
///   "pattern": "*.jpg",
///   "recursive": false
/// }
///
/// // Mkdir operation
/// {
///   "operation": "mkdir",
///   "path": "/path/to/new/directory",
///   "createParents": true
/// }
/// ```
class FileSystemWorker: IosWorker {

    func doWork(input: String?, env: KMPWorkManager.WorkerEnvironment) async throws -> WorkerResult {
        // Register background task to request extra execution time
        // iOS will freeze the app shortly after moving to background otherwise.
        var bgTaskId = UIBackgroundTaskIdentifier.invalid
        bgTaskId = UIApplication.shared.beginBackgroundTask(withName: "BrewkitsFileSystem") {
            NativeLogger.d("FileSystemWorker: Background time expired — ending task")
            UIApplication.shared.endBackgroundTask(bgTaskId)
        }

        defer {
            UIApplication.shared.endBackgroundTask(bgTaskId)
        }

        guard let input = input, !input.isEmpty else {
            return .failure(message: "Input JSON is required")
        }

        // Parse configuration
        guard let data = input.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let operation = json["operation"] as? String else {
            return .failure(message: "Invalid JSON configuration")
        }

        switch operation {
        case "copy":
            return handleCopy(json: json)
        case "move":
            return handleMove(json: json)
        case "delete":
            return handleDelete(json: json)
        case "list":
            return handleList(json: json)
        case "mkdir":
            return handleMkdir(json: json)
        default:
            return .failure(message: "Unknown operation: \(operation)")
        }
    }

    private func handleCopy(json: [String: Any]) -> WorkerResult {
        guard let sourcePath = json["sourcePath"] as? String,
              let destinationPath = json["destinationPath"] as? String else {
            return .failure(message: "Missing sourcePath or destinationPath")
        }

        let overwrite = json["overwrite"] as? Bool ?? false
        let recursive = json["recursive"] as? Bool ?? true

        // Security validation
        do {
            if !SecurityValidator.validateFilePath(sourcePath) {
                return .failure(message: "Invalid source path: \(sourcePath)")
            }
            if !SecurityValidator.validateFilePath(destinationPath) {
                return .failure(message: "Invalid destination path: \(destinationPath)")
            }
        } catch {
            return .failure(message: "Security validation failed: \(error.localizedDescription)")
        }

        let sourceURL = URL(fileURLWithPath: sourcePath)
        let destURL = URL(fileURLWithPath: destinationPath)
        let fileManager = FileManager.default

        // Check source exists
        guard fileManager.fileExists(atPath: sourceURL.path) else {
            return .failure(message: "Source not found: \(sourcePath)")
        }

        // Check destination
        if fileManager.fileExists(atPath: destURL.path) && !overwrite {
            return .failure(message: "Destination already exists: \(destinationPath) (set overwrite=true to replace)")
        }

        // FS-H-008: removed no-op path-traversal guard (always true); SecurityValidator above is sufficient.

        do {
            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: sourceURL.path, isDirectory: &isDirectory)

            let copiedFiles: [URL]
            if isDirectory.boolValue {
                if !recursive {
                    return .failure(message: "Source is a directory, set recursive=true to copy")
                }
                copiedFiles = try copyDirectory(source: sourceURL, destination: destURL, overwrite: overwrite)
            } else {
                try copyFile(source: sourceURL, destination: destURL, overwrite: overwrite)
                copiedFiles = [destURL]
            }

            let totalSize = copiedFiles.reduce(Int64(0)) { sum, url in
                let attrs = try? fileManager.attributesOfItem(atPath: url.path)
                let size = attrs?[.size] as? Int64 ?? 0
                return sum + size
            }

            return .success(
                message: "Copied \(copiedFiles.count) file(s)",
                data: [
                    "operation": "copy",
                    "sourcePath": sourcePath,
                    "destinationPath": destinationPath,
                    "fileCount": copiedFiles.count,
                    "totalSize": totalSize,
                    "files": copiedFiles.map { $0.path }
                ]
            )
        } catch {
            return .failure(message: "Copy failed: \(error.localizedDescription)")
        }
    }

    private func handleMove(json: [String: Any]) -> WorkerResult {
        guard let sourcePath = json["sourcePath"] as? String,
              let destinationPath = json["destinationPath"] as? String else {
            return .failure(message: "Missing sourcePath or destinationPath")
        }

        let overwrite = json["overwrite"] as? Bool ?? false

        // Security validation
        do {
            if !SecurityValidator.validateFilePath(sourcePath) {
                return .failure(message: "Invalid source path: \(sourcePath)")
            }
            if !SecurityValidator.validateFilePath(destinationPath) {
                return .failure(message: "Invalid destination path: \(destinationPath)")
            }
        } catch {
            return .failure(message: "Security validation failed: \(error.localizedDescription)")
        }

        let sourceURL = URL(fileURLWithPath: sourcePath)
        let destURL = URL(fileURLWithPath: destinationPath)
        let fileManager = FileManager.default

        // Check source exists
        guard fileManager.fileExists(atPath: sourceURL.path) else {
            return .failure(message: "Source not found: \(sourcePath)")
        }

        // Check destination
        if fileManager.fileExists(atPath: destURL.path) && !overwrite {
            return .failure(message: "Destination already exists: \(destinationPath) (set overwrite=true to replace)")
        }

        // FS-H-008: removed no-op path-traversal guard (always true); SecurityValidator above is sufficient.

        do {
            // Create parent directory
            try fileManager.createDirectory(
                at: destURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )

            // FS-M-007: rename existing dest to backup before move to prevent data loss.
            // If moveItem later fails, restore the backup so no data is lost.
            var backupURL: URL? = nil
            if fileManager.fileExists(atPath: destURL.path) && overwrite {
                let temp = destURL.deletingLastPathComponent()
                    .appendingPathComponent(".\(destURL.lastPathComponent).bak")
                try fileManager.moveItem(at: destURL, to: temp)
                backupURL = temp
            }

            do {
                // Move file/directory
                try fileManager.moveItem(at: sourceURL, to: destURL)
                // Success — discard backup
                if let backup = backupURL { try? fileManager.removeItem(at: backup) }
            } catch {
                // Restore backup to prevent destination data loss
                if let backup = backupURL {
                    try? fileManager.moveItem(at: backup, to: destURL)
                }
                throw error
            }

            // Count files
            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: destURL.path, isDirectory: &isDirectory)

            let fileCount: Int
            if isDirectory.boolValue {
                fileCount = try countFiles(in: destURL)
            } else {
                fileCount = 1
            }

            return .success(
                message: "Moved \(fileCount) file(s)",
                data: [
                    "operation": "move",
                    "sourcePath": sourcePath,
                    "destinationPath": destinationPath,
                    "fileCount": fileCount
                ]
            )
        } catch {
            return .failure(message: "Move failed: \(error.localizedDescription)")
        }
    }

    private func handleDelete(json: [String: Any]) -> WorkerResult {
        guard let path = json["path"] as? String else {
            return .failure(message: "Missing path")
        }

        let recursive = json["recursive"] as? Bool ?? false

        // Security validation
        do {
            if !SecurityValidator.validateFilePath(path) {
                return .failure(message: "Invalid path: \(path)")
            }
        } catch {
            return .failure(message: "Security validation failed: \(error.localizedDescription)")
        }

        let fileURL = URL(fileURLWithPath: path)
        let fileManager = FileManager.default

        guard fileManager.fileExists(atPath: fileURL.path) else {
            return .failure(message: "Path not found: \(path)")
        }

        // Safety check: prevent accidental deletion of important directories
        let dangerousPaths = ["/", "/System", "/Library", "/usr", "/var"]
        for dangerousPath in dangerousPaths {
            if fileURL.path.hasPrefix(dangerousPath) && fileURL.path.count <= dangerousPath.count + 1 {
                return .failure(message: "Cannot delete protected path: \(path)")
            }
        }

        do {
            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: fileURL.path, isDirectory: &isDirectory)

            let fileCount: Int
            if isDirectory.boolValue {
                if !recursive {
                    return .failure(message: "Path is a directory, set recursive=true to delete")
                }
                fileCount = try countFiles(in: fileURL)
            } else {
                fileCount = 1
            }

            try fileManager.removeItem(at: fileURL)

            return .success(
                message: "Deleted \(fileCount) file(s)",
                data: [
                    "operation": "delete",
                    "path": path,
                    "fileCount": fileCount
                ]
            )
        } catch {
            return .failure(message: "Delete failed: \(error.localizedDescription)")
        }
    }

    private func handleList(json: [String: Any]) -> WorkerResult {
        guard let path = json["path"] as? String else {
            return .failure(message: "Missing path")
        }

        let pattern = json["pattern"] as? String
        let recursive = json["recursive"] as? Bool ?? false

        // Security validation
        do {
            if !SecurityValidator.validateFilePath(path) {
                return .failure(message: "Invalid path: \(path)")
            }
        } catch {
            return .failure(message: "Security validation failed: \(error.localizedDescription)")
        }

        let dirURL = URL(fileURLWithPath: path)
        let fileManager = FileManager.default

        guard fileManager.fileExists(atPath: dirURL.path) else {
            return .failure(message: "Path not found: \(path)")
        }

        var isDirectory: ObjCBool = false
        fileManager.fileExists(atPath: dirURL.path, isDirectory: &isDirectory)

        guard isDirectory.boolValue else {
            return .failure(message: "Path is not a directory: \(path)")
        }

        do {
            // Use streaming enumerator instead of loading all URLs into array
            var fileInfos: [[String: Any]] = []
            var totalSize: Int64 = 0
            var count = 0
            
            let resourceKeys: [URLResourceKey] = [.isDirectoryKey, .fileSizeKey, .contentModificationDateKey]
            var options: FileManager.DirectoryEnumerationOptions = [.skipsHiddenFiles]
            if !recursive {
                options.insert(.skipsSubdirectoryDescendants)
            }
            
            guard let enumerator = fileManager.enumerator(
                at: dirURL,
                includingPropertiesForKeys: resourceKeys,
                options: options
            ) else {
                return .failure(message: "Cannot enumerate directory")
            }
            
            // Fix Regex Injection — pre-compile and safely escape pattern
            let regex = try pattern.flatMap { p -> NSRegularExpression? in
                let escaped = NSRegularExpression.escapedPattern(for: p)
                               .replacingOccurrences(of: "\\*", with: ".*")
                               .replacingOccurrences(of: "\\?", with: ".")
                return try NSRegularExpression(pattern: "^" + escaped + "$", options: .caseInsensitive)
            }

            for case let fileURL as URL in enumerator {
                // IPC Stability: Limit result size to 1000 items
                if count >= 1000 { break }
                
                let vals = try fileURL.resourceValues(forKeys: Set(resourceKeys))
                if vals.isDirectory == true { continue }
                
                let fileName = fileURL.lastPathComponent
                if let r = regex {
                    let range = NSRange(fileName.startIndex..<fileName.endIndex, in: fileName)
                    if r.firstMatch(in: fileName, range: range) == nil { continue }
                }
                
                let size = Int64(vals.fileSize ?? 0)
                fileInfos.append([
                    "path": fileURL.path,
                    "name": fileName,
                    "size": size,
                    "lastModified": vals.contentModificationDate?.timeIntervalSince1970 ?? 0,
                    "isDirectory": false
                ])
                totalSize += size
                count += 1
            }

            return .success(
                message: "Found \(count) file(s)",
                data: [
                    "operation": "list",
                    "path": path,
                    "pattern": pattern ?? "",
                    "recursive": recursive,
                    "fileCount": count,
                    "totalSize": totalSize,
                    "entries": fileInfos.map { $0["path"] as? String ?? "" }
                ]
            )
        } catch {
            return .failure(message: "List failed: \(error.localizedDescription)")
        }
    }

    private func handleMkdir(json: [String: Any]) -> WorkerResult {
        guard let path = json["path"] as? String else {
            return .failure(message: "Missing path")
        }

        let createParents = json["createParents"] as? Bool ?? true

        // Security validation
        do {
            if !SecurityValidator.validateFilePath(path) {
                return .failure(message: "Invalid path: \(path)")
            }
        } catch {
            return .failure(message: "Security validation failed: \(error.localizedDescription)")
        }

        let dirURL = URL(fileURLWithPath: path)
        let fileManager = FileManager.default

        if fileManager.fileExists(atPath: dirURL.path) {
            var isDirectory: ObjCBool = false
            fileManager.fileExists(atPath: dirURL.path, isDirectory: &isDirectory)

            if isDirectory.boolValue {
                return .success(
                    message: "Directory already exists",
                    data: [
                        "operation": "mkdir",
                        "path": path,
                        "created": false
                    ]
                )
            } else {
                return .failure(message: "Path exists but is not a directory: \(path)")
            }
        }

        // FS-H-008: removed no-op path-traversal guard; SecurityValidator above is sufficient.

        do {
            try fileManager.createDirectory(
                at: dirURL,
                withIntermediateDirectories: createParents
            )

            return .success(
                message: "Directory created",
                data: [
                    "operation": "mkdir",
                    "path": path,
                    "created": true
                ]
            )
        } catch {
            return .failure(message: "Mkdir failed: \(error.localizedDescription)")
        }
    }

    // MARK: - Helper Methods

    private func copyFile(source: URL, destination: URL, overwrite: Bool) throws {
        let fileManager = FileManager.default

        // Create parent directory
        try fileManager.createDirectory(
            at: destination.deletingLastPathComponent(),
            withIntermediateDirectories: true
        )

        // Delete existing if overwriting
        if fileManager.fileExists(atPath: destination.path) && overwrite {
            try fileManager.removeItem(at: destination)
        }

        try fileManager.copyItem(at: source, to: destination)
    }

    private func copyDirectory(source: URL, destination: URL, overwrite: Bool) throws -> [URL] {
        let fileManager = FileManager.default
        var copiedFiles: [URL] = []

        // Create destination directory
        try fileManager.createDirectory(at: destination, withIntermediateDirectories: true)

        // Get contents
        guard let enumerator = fileManager.enumerator(
            at: source,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: []
        ) else {
            throw NSError(domain: "FileSystemWorker", code: -1, userInfo: [NSLocalizedDescriptionKey: "Cannot enumerate directory"])
        }

        for case let fileURL as URL in enumerator {
            // FS-M-006: use dropFirst with explicit separator to avoid fragile
            // replacingOccurrences when source.path already ends with "/".
            let sourcePath = source.path.hasSuffix("/") ? source.path : source.path + "/"
            guard fileURL.path.hasPrefix(sourcePath) else { continue }
            let relativePath = String(fileURL.path.dropFirst(sourcePath.count))
            let destURL = destination.appendingPathComponent(relativePath)

            let resourceValues = try fileURL.resourceValues(forKeys: [.isDirectoryKey])
            if resourceValues.isDirectory == true {
                try fileManager.createDirectory(at: destURL, withIntermediateDirectories: true)
            } else {
                try copyFile(source: fileURL, destination: destURL, overwrite: overwrite)
                copiedFiles.append(destURL)
            }
        }

        return copiedFiles
    }

    private func countFiles(in directory: URL) throws -> Int {
        let fileManager = FileManager.default
        var count = 0

        guard let enumerator = fileManager.enumerator(
            at: directory,
            includingPropertiesForKeys: [.isDirectoryKey],
            options: []
        ) else {
            return 0
        }

        for case let fileURL as URL in enumerator {
            let resourceValues = try fileURL.resourceValues(forKeys: [.isDirectoryKey])
            if resourceValues.isDirectory != true {
                count += 1
            }
        }

        return count
    }

}
// FS-L-001: Removed dead code — listFilesRecursive and filterFiles were never called.
